src/lib/allocation_functor/make.ml

```diff
@@ -0,0 +1,258 @@
+open Core_kernel
+
+module Partial = struct
+  module Bin_io (M : Intf.Input.Bin_io_intf) :
+    Intf.Partial.Bin_io_intf with type t := M.t = struct
+    let bin_size_t = M.bin_size_t
+
+    let bin_write_t = M.bin_write_t
+
+    let bin_read_t buf ~pos_ref =
+      Table.attach_finalizer M.id (M.bin_read_t buf ~pos_ref)
+
+    let __bin_read_t__ buf ~pos_ref i =
+      Table.attach_finalizer M.id (M.__bin_read_t__ buf ~pos_ref i)
+
+    let bin_shape_t = M.bin_shape_t
+
+    let bin_writer_t = M.bin_writer_t
+
+    (* TODO: does these last two need wrapped as well? *)
+    let bin_reader_t = M.bin_reader_t
+
+    let bin_t = M.bin_t
```

Comment:
As mentioned on discord, we should just bundle these up as needed:
```ocaml
  let bin_reader_t = { read = bin_read_t; vtag_read = __bin_read_t__ }
  let bin_t = { shape = bin_shape_t; writer = bin_writer_t; reader = bin_reader_t }
```

The `Bin_prot.Utils.Of_minimal` functor should also do this for you, if you'd prefer that route.